### PR TITLE
File hashing 4279 v1

### PR DESCRIPTION
**Pull request by rtkjweeks**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
-
-
-

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: rust/src/ffi/hashing.rs [hunk 1]
```diff
@@ -214,9 +219,19 @@ pub unsafe extern "C" fn SCMd5HashBufferToHex(
 // Functions that are generic over Digest. For the most part the C bindings are
 // just wrappers around these.
 
-unsafe fn update<D: Digest>(digest: &mut D, bytes: *const u8, len: u32) {
+unsafe fn update<D: Digest>(digest: &mut D, bytes: *const u8, mut len: u32, bytes_hashed: *mut u32) {
+    if g_hash_byte_limit > 0 && !bytes_hashed.is_null() {
+	let bytes_remaining = g_hash_byte_limit - *bytes_hashed;
```
Comment (victorjulien) [inline rust/src/ffi/hashing.rs:224]: indent is off, should be 4 spaces

File: rust/src/ffi/hashing.rs [hunk 2]
```diff
@@ -214,9 +219,19 @@ pub unsafe extern "C" fn SCMd5HashBufferToHex(
...
+	if len > bytes_remaining {
+	    len = bytes_remaining;
+	}
+
+	*bytes_hashed += len
+    }
+
     let data = std::slice::from_raw_parts(bytes, len as usize);
     digest.update(data);
+
```
Comment (victorjulien) [inline rust/src/ffi/hashing.rs:234]: unnecessary newline addition

File: rust/src/ffi/hashing.rs [hunk 3]
```diff
@@ -32,15 +32,20 @@ pub const SC_MD5_HEX_LEN: usize = 32;
 // with .0.
 pub struct SCSha256(Sha256);
 
+extern "C" {
+    #[no_mangle]
+    static g_hash_byte_limit: u32;
```
Comment (victorjulien) [inline rust/src/ffi/hashing.rs:37]: is this essentially an `extern uint32_t g_hash_byte_limit` referring to the C variable? Seems newer Rust doesn't like it, see the Fedora CI failures

File: src/util-file.h [hunk 4]
```diff
@@ -85,10 +85,13 @@ typedef struct File_ {
 #endif
     struct File_ *next;
     SCMd5 *md5_ctx;
+    uint32_t md5_num_bytes;
```
Comment (victorjulien) [inline src/util-file.h:88]: I'm not too excited about having to grow this structure like this. Is there another way we can do this?

If not, at least the in-memory layout needs to be improved by avoiding holes in the memory layout (hint: check `pahole`)

File: src/util-file.h [hunk 5]
```diff
@@ -141,7 +144,7 @@ int FileOpenFileWithId(FileContainer *, const StreamingBufferConfig *,
         const uint8_t *data, uint32_t data_len, uint16_t flags);
 
 /**
- *  \brief Close a File
+ *  \brief Close all Files in the container
```
Comment (victorjulien) [inline src/util-file.h:147]: unrelated change, should be in its own commit

File: src/util-file.h [hunk 6]
```diff
@@ -158,6 +161,7 @@ int FileCloseFileById(FileContainer *, uint32_t track_id,
 int FileCloseFilePtr(File *ff, const uint8_t *data,
         uint32_t data_len, uint16_t flags);
 
+
```
Comment (victorjulien) [inline src/util-file.h:164]: unnecessary newline change

File: src/util-file.c [hunk 7]
```diff
@@ -931,6 +955,7 @@ int FileOpenFileWithId(FileContainer *ffc, const StreamingBufferConfig *sbcfg,
 
 int FileCloseFilePtr(File *ff, const uint8_t *data,
         uint32_t data_len, uint16_t flags)
+
```
Comment (victorjulien) [inline src/util-file.c:958]: unneeded newline

**Review by victorjulien — CHANGES_REQUESTED** [review]
Can you create matching suricata-verify tests and refer to them in your next PR: `suricata-verify-pr: XXX` (where XXX is just the PR number)

Commits need cleanup, please check the existing commits in our tree. We generally use `subsys: subject` as the subject line.

Please make sure all committers in the PRs have signed the CLA.

Also, please check CI failures. Only CIFuzz is a known issue right now so ignore that.

Thanks!

File: src/util-file.h (hunk 4 above)
Comment (rtkjweeks) [reply src/util-file.h]: Couldn't find a reasonable way to remove these without removing functionality (ie., only using one hash function, not allowing multiple hashing functions).
SHA1 is only 20 bytes, so there's room to squeeze one in there; the others I added after, so they were at least co-located.

**Comment by rtkjweeks** [comment]
Closing per V2 PR: https://github.com/OISF/suricata/pull/5846
